### DISCUSSION TITLE
Better Clojure highlighting with nextjournal/clojure-mode

### DISCUSSION THREAD

**Discussion by borkdude** [borkdude/blog#4] [Posts]
https://blog.michielborkent.nl/better-clojure-highlighting.html

**Comment by andreyorst** [comment]
Thanks for this post! I've enjoyed reading both this one and the [next one](https://blog.michielborkent.nl/writing-clojure-highlighter.html) on this topic as well. However, I have some notes.

It may sound a bit strange, but why is this better? There's not much of a difference from highlight.js at the first glance, and looking at the code I see almost everything in purple-ish color, and as a result, color has almost no meaning now because *everything* is purple, blue, and dark-blue. I mean, syntax highlighting is supposed to highlight something important - if we look into the Wiktionary definition of [highlight](https://en.wiktionary.org/wiki/highlight) one of its meanings is: "especially significant or interesting detail or event". However, the code is mostly the same color.

What I like about Emacs default approach to highlighting (for many languages, not just Clojure) is that it tries to highlight only important things. A function call is not important, function definition - is. Macro call, on the other hand, is important, because there's no other way to differentiate what is a macro (besides indentation, but I'm talking only about syntax highlighting here). A dynamically scoped variable is important, because it may change depending on the binding scope, an ordinary variable - isn't. Here's how Emacs highlights the code from your article:

![image](https://user-images.githubusercontent.com/19470159/140737193-cf55798b-b511-4a82-8764-9dc28ea775e9.png)

And here's the highlighting from your blog:

![image](https://user-images.githubusercontent.com/19470159/140733329-45824655-1892-4de9-8a2e-fe99f595f90d.png)

So let me list the key differences I find important.

In Emacs, all special symbols have their own, distinct, color: `defn`, `let`, `new`, `fn`. In your blog, they're colored the same as function calls, even though they're not semantically the same as *function* calls. 

Emacs highlights the namespace part in a different color because namespaces indicate that the symbol comes from another package, which is important. In your implementation, they're the same color as a function call. 

Looking further, even some ordinary symbol names are being highlighted with the same color that is used for function calls in your implementation, for instance, the `EditorState` or `EditorView` are. Other bindings on the other hand are highlighted with a far more different cyan-ish color, but these aren't as significant as, say, special forms, or dynamic variables, why even highlight them?

Don't get me wrong, syntax highlighting is both hard and personal. I just tried to highlight<sup>(see what I did there?)</sup> the potential problems I see in this implementation. I like the minimalist approach, where only significant parts are highlighted because I find it the most practical one. Others may have different views on this, and it's fine. There's a quite old [blog post against syntax highlighting](https://www.linusakesson.net/programming/syntaxhighlighting/) which spanned a lot of somewhat interesting discussions, so it's fair to say that there are as many opinions as there are people.

I should say, that in my personal practice I went from full rainbow mode to zero highlighting over the years, and finally found the perfect spot at what Emacs does by default - highlight only interesting parts, leave everything else the default color. This leaves a lot of space for what colors to use because we're often limited by 16-20 colors before it becomes a color mess. So the fewer things you need to highlight, the more colors you can potentially utilize without introducing clutter, or losing meaning.

**Reply by borkdude** [reply]
All very good points. I went a little bit overboard the current styling. Special symbols may need special highlighting. I think the current implementation can be tweaked quite easily.

I just made another thread for the most recent post. 

https://github.com/borkdude/blog/discussions/11
